File: harnesses/harness.js
const enc = new TextEncoder();
const dec = new TextDecoder();

const elog = (...a) => {
    const w = tjs.stderr.getWriter();
    w.write(enc.encode(a.map(String).join(' ') + '\n'));
    w.releaseLock();
};

console.log = elog;
console.warn = elog;
console.info = elog;
console.debug = elog;

const reader = tjs.stdin.getReader();
const chunks = [];
for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    chunks.push(value);
}

const concat = (parts) => {
    let len = 0;
    for (const c of parts) len += c.length;
    const out = new Uint8Array(len);
    let off = 0;
    for (const c of parts) { out.set(c, off); off += c.length; }
    return out;
};

const input = JSON.parse(dec.decode(concat(chunks)));

const dir = new URL('.', import.meta.url).pathname;
const { run } = await import(dir + 'transform.js');
const out = await run(input);

const writer = tjs.stdout.getWriter();
await writer.write(enc.encode(JSON.stringify(out)));
writer.releaseLock();

File: harnesses/harness.py
import sys
import json
import os
import importlib.util

stdout = os.dup(1)
os.dup2(2, 1)
sys.stdout = sys.stderr

input = json.loads(sys.stdin.buffer.read())

spec = importlib.util.spec_from_file_location(
    "transform",
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "transform.py"),
)

if spec is None or spec.loader is None:
    raise Exception("Invalid module transform.py")

mod = importlib.util.module_from_spec(spec)
spec.loader.exec_module(mod)

output = mod.run(input)

os.write(stdout, json.dumps(output).encode())
os.close(stdout)
